### ISSUE TITLE
How do I access SELinux Python bindings?

### ISSUE THREAD

**Issue by unclechu** [NixOS/nixpkgs#70980]
I can't find any related package [here](https://nixos.org/nixos/packages.html?channel=nixos-19.09) (I mean some separated package, bindings-only, like `python3-libselinux` in Fedora).

That's what I'm trying to do:
```
λ nix-shell -p python3 libselinux --run "python3 -c 'import selinux; print(selinux)'"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'selinux'
```

These are probably related:
- https://github.com/NixOS/nixpkgs/issues/17808
- https://github.com/NixOS/nixpkgs/pull/17817

But as far as I understand that pull request solves it only for NixOS but not for other OSes which uses Nix, did I get it right?

_P.S. I'm using `nix` under `Fedora Workstation 30._

**Comment by teto** [comment]
There is a package called `selinux-python`, is it what you are looking for ?
I tried to `import selinux` with it but it failed so maybe not, or maybe it needs some tweaks, I would expect such a package to live in python-packages.nix instead of all-packages.nix

**Comment by unclechu** [comment]
@teto Nope, `selinux-python` package doesn't provide `selinux` python package, you can see:

`$PWD`: `/nix/store/pjzm8wikvsfkyr3y8qhqncq9w9ard77f-selinux-python-2.7/lib/python3.7/site-packages`

```
$ ls -lah
total 112K
dr-xr-xr-x. 4 uc uc   92  1. 1.  1970 .
dr-xr-xr-x. 3 uc uc   27  1. 1.  1970 ..
-r-xr-xr-x. 1 uc uc 101K  1. 1.  1970 seobject.py
dr-xr-xr-x. 2 uc uc 4,0K  1. 1.  1970 sepolgen
dr-xr-xr-x. 5 uc uc  263  1. 1.  1970 sepolicy
-r--r--r--. 1 uc uc  227  1. 1.  1970 sepolicy-1.1-py3.7.egg-info
```

**Comment by unclechu** [comment]
As it turns out from https://github.com/NixOS/nixpkgs/issues/17808 I could explicitly enable `enablePython` flag, like this:
```nix
pkgs.libselinux.override { enablePython = true; }
```
Also you can find this flag [here](https://github.com/NixOS/nixpkgs/blob/076860e0340a5e4a909b9a710e186508b14d1c90/pkgs/os-specific/linux/libselinux/default.nix) (please note it is supposed to be turned on by default), but it seems when I turn it on it doesn't do the thing I'd expect it to do. At least this gives me nothing (no python packages provided by `libselinux` nix package):
```bash
printf '%s\n' /nix/store/*libselinux* | perl -nE 'chomp; say if -d' | xargs -i find {} -name '*python*'
```

**Comment by jtojnar** [comment]
`libselinux` build Python 2 bindings by default for some reason. We should probably add something like

```
--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -2537,6 +2537,11 @@ in {
 
   latexcodec = callPackage ../development/python-modules/latexcodec {};
 
+  libselinux = (toPythonModule (pkgs.libselinux.override {
+    enablePython = true;
+    inherit python;
+  })).py;
+
   libsexy = callPackage ../development/python-modules/libsexy {
     inherit (pkgs) libsexy pkgconfig;
   };
```

so that users can use `nix-shell -p python3 python3.pkgs.libselinux --run "python3 -c 'import selinux; print(selinux)'"`.

I am not sure how to make it work with `buildEnv`, though, so `nix-shell -E '(with import <nixpkgs> {}; python3.withPackages (pp: [ pp.libselinux ]))' --run "python3 -c 'import selinux; print(selinux)'"`  will still fail.

**Comment by jtojnar** [comment]
This monstrosity seems to work the best:

```nix
  libselinux = (((toPythonModule pkgs.libselinux).overrideAttrs (super: {
    meta = super.meta // {
      outputsToInstall = [ "py" ];
    };
  })).override {
    enablePython = true;
    inherit python;
  }).py
```

* `toPythonModule` for `buildEnv` (`python.withPackages`) to be able to recognize it as Python module.
* `inherit python` to use the correct Python version.
* `.py` to make `python3.pkgs.libselinux` work.
* Changing `outputsToInstall` does not appear to be necessary but it might still be nice to not pollute the env with the selinux manpage etc.
* Order of operations should not matter, except that `.py` needs to be the last step.

Also `python3.withPackages` does not seem to work with `nix-shell` any more for some reason, so `env $(nix-build -E '(with import <nixpkgs> {}; python3.withPackages (pp: [ (pp.libselinux) ]))')/bin/python3 -c 'import selinux; print(selinux)` or `nix run '(with import <nixpkgs> {}; python3.withPackages (pp: [ pp.libselinux ]))' -c python -c 'import selinux; print(selinux)'` needs to be used instead. That works with the first patch as well.

**Comment by unclechu** [comment]
@jtojnar 

For both:
```bash
env $(nix-build -E '(with import <nixpkgs> {}; python3.withPackages (pp: [ (pp.libselinux) ]))')/bin/python3 -c 'import selinux; print(selinux)'
```
and 
```bash
nix run '(with import <nixpkgs> {}; python3.withPackages (pp: [ pp.libselinux ]))' -c python -c 'import selinux; print(selinux)'
```
I get: `error: attribute 'libselinux' missing, at (string):1:57`

**Comment by unclechu** [comment]
By writing this:
```nix
  libselinuxWithPythonBindings =
    (((pkgs.pythonPackages.toPythonModule pkgs.libselinux).overrideAttrs (super: {
      meta = super.meta // {
        outputsToInstall = [ "py" ];
      };
    })).override {
      enablePython = true;
      inherit (pkgs.python);
    }).py;
```
It seems I've got some python bindings but for python2:
```
# /nix/store/jxr1h4bf7mlyrqsai5j0i635dndi6x2w-libselinux-2.9-py/lib/python2.7/site-packages/selinux
λ ls
audit2why.so  __init__.py
```

**Comment by unclechu** [comment]
It seems I fixed it by overriding `python` for `pkgs.libselinux`:
```
λ which python3
/nix/store/swy0p01xr0wyh907d67hkxr1g0kngcpn-python3-3.7.4/bin/python3
λ python3
Python 3.7.4 (default, Jul  8 2019, 18:31:06)
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import selinux
>>> selinux
<module 'selinux' from '/nix/store/y2qfmd4yyx3jp0zbqiwh7x3mjnz6k3hg-libselinux-2.9-py/lib/python3.7/site-packages/selinux/__init__.py'>
```
This is what I have added to my `shell.nix`:
```nix
  libselinuxPythonBindings =
    (((pkgs.python37Packages.toPythonModule (pkgs.libselinux.override {
      python = pkgs.python37;
    })).overrideAttrs (super: {
      meta = super.meta // {
        outputsToInstall = [ "py" ];
      };
    })).override {
      enablePython = true;
    }).py;
  pythonWithSelinuxBindings = pkgs.python37.withPackages (_: [ libselinuxPythonBindings ]);
```
And added `pythonWithSelinuxBindings` to the `buildInputs`.

**Comment by jtojnar** [comment]
Sorry, I should have been more explicit, the override should go to `pkgs/top-level/python-packages.nix`, see https://github.com/NixOS/nixpkgs/pull/71659.

**Comment by paluh** [comment]
It seems somewhat relevant so I'm posting it here. It seems that `libselinux.py` no longer provides python2.7 `selinux` module in case for example of [`nixpart0`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/filesystems/nixpart/0.4/default.nix#L11) and its `blivet` module. Minimal example which breaks:

```nix
let
  pkgs = import ./default.nix {};
in
  pkgs.stdenv.mkDerivation {
    name = "debug-";
    NIX_PATH="nixpkgs=${pkgs.path}";
    buildInputs = [
      # pkgs.python2Packages.libselinux
      pkgs.python2Packages.nixpart0
    ];
    shellHook = ''nixpart --help'';
  }
```

It is enough to uncomment `pkgs.python2Packages.libselinux` line to get a working `nixpart`.Tested against fresh `nixpkgs/master`.

**Comment by jtojnar** [comment]
This issue was fixed by https://github.com/NixOS/nixpkgs/pull/71659. You can just use `pythonX.pkgs.libselinux` to get the bindings.

The `nixpart0` failure is unrelated (caused by 56e557257df675bc6b2cbebf3ba5bc6d51b2321c) and I fixed it in 07b8ab4ff79. Thanks for pointing it out.

**Comment by paluh** [comment]
Cool. Thanks for quick feedback and this fix!
